semistruct_test.go

```diff
@@ -0,0 +1,147 @@
+package semistruct
+
+import (
+	"github.com/leanovate/gopter"
+	"github.com/leanovate/gopter/gen"
+	"github.com/leanovate/gopter/prop"
+	"testing"
+)
+
+type testpair struct {
+	logline string
+}
+
+// Hand-written unit tests for common cases this human can think of.
+var tests = []testpair{
+	{`!< 2 [cl7323:featstore:sess_fun] { ONE=two DOS="wah=hh-77" } >!`},
+	{`!< 7 [local:pktproc:parsePkt] { PROTO=icmp ERROR="unassigned_type" ACTION=dropped } >!`},
+	{`!< 4 [cl610:featstore:sess_fun] { UNKNOWN_SESS=3e0a5ae3-7c2f-454e-828c-d838a18d5d8e } >!`},
+	{`!< 0 [cl2] { UNSTRUCT_MSG="some random debugging spam" } >!`},
+	{`!< 6 [cl2:filestore:notify] { FILE="cas%20AV_!AA%20nvvpa.jpg" } >!`},
+	{`!< 0 [cl2:filestore:notify] { FILE2="some blah.jpg" } >!`},
+	{`!< 2 [cl2:filestore:notify] { FILE2="some blah.jpg" } >!`},
+	{`!< 2 [cl2:filestore:notify] { FILE2="some bl\ah.jpg" } >!`},
+	{`!< 2 [cl2:filestore:notify] { FILE2="some bl|ah.jpg" } >! ** (process:4707): WARNING **: send_infos_cb: No such interface 'org.gtk.vfs.Enumerator' on object at path /org/gtk/vfs/client/enumerator/385 (g-dbus-error-quark, 19)`},
+	{`!< 2 { FILE2="some blah.jpg" } >!`},
+	{`!< 2 [cl2:filestore:notify] >!`},
+	{`!< 2 >!`},
+	{"!< 3 [blah] { FLFANHB2X6UBMERR=\"kL]_:;\" } >!"},
+}
+
+// Iterate over the hand-written tests and attempt to parse each line.
+func TestParser(t *testing.T) {
+	p := ParseSemistruct()
+
+	for _, pair := range tests {
+		res, _ := p.ParseString(pair.logline)
```

Comment:
Same comment I left in the docker repo:

The reason why is because I didn't fully understand what that pattern actually was until you brought it up - I've read the documentation on the `ok` / `error` pattern now. The `semistruct_line` will be `nil` if the parse fails but I looked at the function definition for `ParseString` and it can either return `nil` in the second return value or an error if there is a failure.

I'll make sure I go through all of the code to make sure I handle errors properly.